File: frontend/src/lib/validations/notificationConfig.ts
import { z } from 'zod';

// Notification Mode enum (0 = DELAY, 1 = FIXED_TIME)
export const NotificationModeEnum = z.union([z.literal(0), z.literal(1)]);
export type NotificationMode = z.infer<typeof NotificationModeEnum>;

// Repeat Policy enum (0 = NO_REPEAT, 2 = DAILY)
export const RepeatPolicyEnum = z.union([z.literal(0), z.literal(2)]);
export type RepeatPolicy = z.infer<typeof RepeatPolicyEnum>;

// Scheduling Mode enum (0 = RANDOM, 1 = SEQUENTIAL)
export const SchedulingModeEnum = z.union([z.literal(0), z.literal(1)]);
export type SchedulingMode = z.infer<typeof SchedulingModeEnum>;

// Notification Message schema
export const notificationMessageSchema = z.object({
  title: z.string().min(1, 'Title is required'),
  body: z.string().min(1, 'Body is required'),
  payload: z.string().default(''),
  androidChannelId: z.string().default(''),
  iosCategory: z.string().default(''),
  offsetSeconds: z.number().int().min(0).default(0),
});

// Notification Strategy schema
export const notificationStrategySchema = z.object({
  id: z.string().min(1, 'Strategy ID is required'),
  name: z.string().min(1, 'Strategy name is required'),
  mode: NotificationModeEnum,
  delaySeconds: z.number().int().min(0).default(0),
  fixedHour: z.number().int().min(0).max(23).default(0),
  fixedMinute: z.number().int().min(0).max(59).default(0),
  fixedDaysOffset: z.number().int().min(0).default(0),
  repeatPolicy: RepeatPolicyEnum,
  repeatSeconds: z.number().int().min(0).default(0),
  active: z.boolean().default(true),
  autoScheduled: z.boolean().default(true),
  schedulingMode: SchedulingModeEnum,
  defaultChannelId: z.string().min(1, 'Default channel ID is required'),
  notifications: z.array(notificationMessageSchema).min(1, 'At least one notification is required'),
});

// Notification Channel schema
export const notificationChannelSchema = z.object({
  id: z.string().min(1, 'Channel ID is required'),
  name: z.string().min(1, 'Channel name is required'),
  description: z.string().default(''),
  defaultBadge: z.number().int().min(0).default(1),
  importance: z.number().int().min(0).max(5).default(3),
  enableLights: z.boolean().default(true),
  enableVibration: z.boolean().default(true),
  canBypassDnd: z.boolean().default(false),
  canShowBadge: z.boolean().default(true),
  lockScreenVisibility: z.number().int().min(-1).max(1).default(0),
});

// Notification Config schema
export const notificationConfigSchema = z
  .object({
    enable: z.boolean().default(true),
    strategies: z.array(notificationStrategySchema).default([]),
    channels: z.array(notificationChannelSchema).min(1, 'At least one channel is required'),
  })
  .refine(
    (data) => {
      const ids = data.strategies.map((s) => s.id);
      return ids.length === new Set(ids).size;
    },
    {
      message: 'Strategy IDs must be unique',
      path: ['strategies'],
    }
  )
  .refine(
    (data) => {
      const ids = data.channels.map((c) => c.id);
      return ids.length === new Set(ids).size;
    },
    {
      message: 'Channel IDs must be unique',
      path: ['channels'],
    }
  );

export type NotificationMessage = z.infer<typeof notificationMessageSchema>;
export type NotificationStrategy = z.infer<typeof notificationStrategySchema>;
export type NotificationChannel = z.infer<typeof notificationChannelSchema>;
export type NotificationConfig = z.infer<typeof notificationConfigSchema>;

// Default notification message
export const defaultMessage: NotificationMessage = {
  title: '',
  body: '',
  payload: '',
  androidChannelId: '',
  iosCategory: '',
  offsetSeconds: 0,
};

// Default notification channel
export const defaultChannel: NotificationChannel = {
  id: '',
  name: '',
  description: '',
  defaultBadge: 0,
  importance: 0,
  enableLights: false,
  enableVibration: false,
  canBypassDnd: false,
  canShowBadge: false,
  lockScreenVisibility: 0,
};

// Default notification strategy
export const defaultStrategy: NotificationStrategy = {
  id: '',
  name: '',
  mode: 0,
  delaySeconds: 0,
  fixedHour: 0,
  fixedMinute: 0,
  fixedDaysOffset: 0,
  repeatPolicy: 0,
  repeatSeconds: 0,
  active: false,
  autoScheduled: false,
  schedulingMode: 0,
  defaultChannelId: '',
  notifications: [],
};

// Default notification config
export const defaultNotificationConfig: NotificationConfig = {
  enable: false,
  strategies: [],
  channels: [],
};

File: backend/app/schemas/config_sections/notification_config.py
from typing import List
from enum import Enum
from pydantic import BaseModel, Field, field_validator, ConfigDict


class NotificationMode(int, Enum):
    """Notification scheduling mode"""
    DELAY = 0  # Delay from last session
    FIXED_TIME = 1  # Fixed time of day


class RepeatPolicy(int, Enum):
    """Notification repeat policy"""
    NO_REPEAT = 0
    DAILY = 2


class SchedulingMode(int, Enum):
    """Notification scheduling mode"""
    RANDOM = 0
    SEQUENTIAL = 1


class NotificationMessage(BaseModel):
    """Individual notification message"""
    model_config = ConfigDict(populate_by_name=True)
    
    title: str = Field(..., min_length=1, description="Notification title key")
    body: str = Field(..., min_length=1, description="Notification body key")
    payload: str = Field(default="", description="Custom payload data")
    androidChannelId: str = Field(default="", description="Android channel override")
    iosCategory: str = Field(default="", description="iOS category")
    offsetSeconds: int = Field(default=0, ge=0, description="Offset from strategy time")


class NotificationStrategy(BaseModel):
    """Notification strategy configuration"""
    model_config = ConfigDict(populate_by_name=True)
    
    id: str = Field(..., min_length=1, description="Unique strategy identifier")
    name: str = Field(..., min_length=1, description="Human-readable name")
    mode: NotificationMode = Field(..., description="Scheduling mode (delay or fixed time)")
    delaySeconds: int = Field(default=0, ge=0, description="Delay in seconds (for DELAY mode)")
    fixedHour: int = Field(default=0, ge=0, le=23, description="Hour for fixed time (0-23)")
    fixedMinute: int = Field(default=0, ge=0, le=59, description="Minute for fixed time (0-59)")
    fixedDaysOffset: int = Field(default=0, ge=0, description="Days offset for fixed time")
    repeatPolicy: RepeatPolicy = Field(..., description="Repeat policy")
    repeatSeconds: int = Field(default=0, ge=0, description="Repeat interval in seconds")
    active: bool = Field(default=True, description="Whether strategy is active")
    autoScheduled: bool = Field(default=True, description="Auto-schedule on trigger")
    schedulingMode: SchedulingMode = Field(..., description="Random or sequential selection")
    defaultChannelId: str = Field(..., min_length=1, description="Default notification channel")
    notifications: List[NotificationMessage] = Field(min_length=1, description="List of notification messages")


class NotificationChannel(BaseModel):
    """Android notification channel configuration"""
    model_config = ConfigDict(populate_by_name=True)
    
    id: str = Field(..., min_length=1, description="Channel identifier")
    name: str = Field(..., min_length=1, description="Channel name")
    description: str = Field(default="", description="Channel description")
    defaultBadge: int = Field(default=1, ge=0, description="Default badge count")
    importance: int = Field(default=3, ge=0, le=5, description="Channel importance (0-5)")
    enableLights: bool = Field(default=True, description="Enable notification lights")
    enableVibration: bool = Field(default=True, description="Enable vibration")
    canBypassDnd: bool = Field(default=False, description="Can bypass Do Not Disturb")
    canShowBadge: bool = Field(default=True, description="Can show badge")
    lockScreenVisibility: int = Field(default=0, ge=-1, le=1, description="Lock screen visibility")


class NotificationConfig(BaseModel):
    """Complete notification configuration"""
    model_config = ConfigDict(populate_by_name=True)
    
    enable: bool = Field(default=True, description="Enable notification system")
    strategies: List[NotificationStrategy] = Field(
        default_factory=list,
        description="Notification strategies"
    )
    channels: List[NotificationChannel] = Field(
        default_factory=list,
        description="Notification channels"
    )
    
    @field_validator('strategies')
    @classmethod
    def validate_unique_strategy_ids(cls, v: List[NotificationStrategy]) -> List[NotificationStrategy]:
        """Ensure strategy IDs are unique"""
        if not v:
            return v
        ids = [s.id for s in v]
        if len(ids) != len(set(ids)):
            raise ValueError('Strategy IDs must be unique')
        return v
    
    @field_validator('channels')
    @classmethod
    def validate_unique_channel_ids(cls, v: List[NotificationChannel]) -> List[NotificationChannel]:
        """Ensure channel IDs are unique"""
        ids = [c.id for c in v]
        if len(ids) != len(set(ids)):
            raise ValueError('Channel IDs must be unique')
        return v
